update SPM doc for SQL plan auto capture (#8619)
pingcap/docs-cn: sql-plan-management.md

CREATE [GLOBAL | SESSION] BINDING FOR BindableStmt USING BindableStmt;
```

该语句可以在 GLOBAL 或者 SESSION 作用域内为 SQL 绑定执行计划。目前，如下 SQL 类型 (BindableStmt) 可创建执行计划绑定：`SELECT`、`DELETE`、`UPDATE` 和带有 `SELECT` 子查询的 `INSERT`/`RELACE`。

> **注意：**
>

> > 绑定的优先级高于手工添加的 Hint，即在有绑定的时候执行带有 Hint 的语句时，该语句中控制优化器行为的 Hint 不会生效，但是，其他类别的 Hint 仍然能够生效。

Reviewer: ```suggestion
> 绑定的优先级高于手工添加的 Hint，即在有绑定的时候执行带有 Hint 的语句时，该语句中控制优化器行为的 Hint 不会生效，但是其他类别的 Hint 仍然能够生效。
```

CREATE [GLOBAL | SESSION] BINDING FOR BindableStmt USING BindableStmt;
```

> 该语句可以在 GLOBAL 或者 SESSION 作用域内为 SQL 绑定执行计划。目前支持的可创建执行计划绑定的 SQL 类型 (BindableStmt) 包括：`SELECT`，`DELETE`，`UPDATE` 和带有 `SELECT` 子查询的 `INSERT`/`RELACE`。
> 
> > **注意：**
> >
> > 绑定的优先级高于手工添加的 Hint，即在有绑定的时候执行带有 Hint 的语句，该语句中控制优化器行为的 Hint 不会生效，但是其他类别的 Hint 仍然能够生效。

Reviewer 2: ```suggestion
该语句可以在 GLOBAL 或者 SESSION 作用域内为 SQL 绑定执行计划。目前，如下 SQL 类型 (BindableStmt) 可创建执行计划绑定：`SELECT`、`DELETE`、`UPDATE` 和带有 `SELECT` 子查询的 `INSERT`/`RELACE`。

> **注意：**
>
> 绑定的优先级高于手工添加的 Hint，即在有绑定的时候执行带有 Hint 的语句时，该语句中控制优化器行为的 Hint 不会生效，但是，其他类别的 Hint 仍然能够生效。
```

>
> 对于 `PREPARE`/`EXECUTE` 语句组，或者用二进制协议执行的查询，创建执行计划绑定的对象应当是查询语句本身，而不是 `PREPARE`/`EXECUTE` 语句。

> **注意：**
>

> > 绑定的优先级高于手工添加的 Hint，即在有绑定的时候执行带有 Hint 的语句，该语句中控制优化器行为的 Hint 不会生效，但是其他类别的 Hint 仍然能够生效。

Reviewer 3: 这个提示能否放在靠近 ## 创建绑定 标题的位置呢？

```sql
-- 删除 session 中创建的 binding
DROP session binding for SELECT * FROM t1, t2 WHERE t1.id = t2.id;

-- 重新查看该 SQL 的执行计划
explain SELECT * FROM t1,t2 WHERE t1.id = t2.id;
```

在这里 SESSION 作用域内被删除掉的绑定会屏蔽 GLOBAL 作用域内相应的绑定，优化器不会为 `SELECT` 语句添加 `sm_join(t1, t2)` hint，`explain` 给出的执行计划中最上层节点并不被 hint 固定为 MergeJoin，而是由优化器经过代价估算后自主进行选择。

> **注意：**
>
> 执行 `DROP GLOBAL BINDING` 会删除当前 tidb-server 实例缓存中的绑定，并将系统表中对应行的状态修改为 'deleted'。该语句不会直接删除系统表中的记录，因为其他 tidb-server 实例需要读取系统表中的 'deleted' 状态来删除其缓存中对应的绑定。对于这些系统表中状态为 'deleted' 的记录，后台线程每隔 100 个 `bind-info-lease`（默认值为 `3s`，合计 `300s`）会触发一次对 `update_time` 在 10 个 `bind-info-lease` 以前的绑定（确保所有 tidb-server 实例已经读取过这个 'deleted' 状态并更新完缓存）的回收清除操作。

### 变更绑定状态

{{< copyable "sql" >}}

```sql
SET BINDING [ENABLED | DISABLED] FOR BindableStmt;
```

该语句可以在 GLOBAL 作用域内变更指定执行计划的绑定状态，默认作用域为 GLOBAL，该作用域不可更改。

> 使用时，只能将 `Disabled` 的绑定改为 `Enabled` 状态，或将 `Enabled` 的绑定改为 `Disabled` 状态。如果没有可以改变状态的绑定，则会输出一条内容为 `There are no bindings can be set the status. Please check the SQL text` 的 warning 警告进行提示。需要注意的是，当绑定被设置成 `Disabled` 状态时，查询语句不会使用该绑定。

Reviewer: ```suggestion
使用时，只能将 `Disabled` 的绑定改为 `Enabled` 状态，或将 `Enabled` 的绑定改为 `Disabled` 状态。如果没有可以改变状态的绑定，则会输出一条内容为 `There are no bindings can be set the status. Please check the SQL text` 的警告。需要注意的是，当绑定被设置成 `Disabled` 状态时，查询语句不会使用该绑定。
```

> 使用时，只能将 `Disabled` 的绑定改为 `Enabled` 状态，或将 `Enabled` 的绑定改为 `Disabled` 状态。如果没有可以改变状态的绑定，则会输出一条警告日志：`The memory usage of all available bindings exceeds the cache's mem quota. As a result, all available bindings cannot be held on the cache. Please increase the system variable 'tidb_mem_quota_binding_cache' and execute 'admin reload bindings' to ensure that all bindings are available normally`。需要注意的是，当绑定被设置成 `Disabled` 状态时，查询语句不会使用该绑定。

Reviewer 4: 这里的警告日志是不是 “货不对板”，实际上应该有两种情况，如果是 cache 内存不足，是你这个错误；
如果没有可以改变状态的绑定，是不是为 `There are no bindings can be set the status. Please check the SQL text`?

>
> 执行 `DROP GLOBAL BINDING` 会删除当前 tidb-server 实例缓存中的绑定，并将系统表中对应行的状态修改为 'deleted'。该语句不会直接删除系统表中的记录，因为其他 tidb-server 实例需要读取系统表中的 'deleted' 状态来删除其缓存中对应的绑定。对于这些系统表中状态为 'deleted' 的记录，后台线程每隔 100 个 `bind-info-lease`（默认值为 `3s`，合计 `300s`）会触发一次对 `update_time` 在 10 个 `bind-info-lease` 以前的绑定（确保所有 tidb-server 实例已经读取过这个 'deleted' 状态并更新完缓存）的回收清除操作。

### 变更绑定状态

{{< copyable "sql" >}}

```sql
SET BINDING [ENABLED | DISABLED] FOR BindableStmt;
```

> 该语句可以在 GLOBAL 作用域内变更指定的执行计划绑定的状态，作用域不可指定，默认作用域为 GLOBAL。

Reviewer 2: ```suggestion
该语句可以在 GLOBAL 作用域内变更指定执行计划绑定的状态为 `Enabled` 或者 `Disabled`，作用域默认为 GLOBAL，不可指定。
```

> 该语句可以在 GLOBAL 作用域内变更指定的执行计划绑定的状态，作用域不可指定，默认作用域为 GLOBAL。

Reviewer 3: ```suggestion
该语句可以在 GLOBAL 作用域内变更指定执行计划的绑定状态，默认作用域为 GLOBAL，该作用域不可更改。
```

> 该语句可以将绑定的状态设置成 `Enabled` 或者 `Disabled` 状态，其中只有 `Disabled` 状态的绑定可以被设置成 `Enabled` 状态，只有 `Enabled` 状态的绑定可以被设置成 `Disabled` 状态。如果当前没有可以改变状态的绑定，则会在日志中输出一条警告日志进行提示。需要注意的是当绑定的状态被设置成 `Disabled` 状态的时候，该绑定不会被查询语句所使用。

Reviewer 3: 会输出什么样的警告日志提示呢？能否提供一个示例呢？

> 该语句可以将绑定的状态设置成 `Enabled` 或者 `Disabled` 状态，其中只有 `Disabled` 状态的绑定可以被设置成 `Enabled` 状态，只有 `Enabled` 状态的绑定可以被设置成 `Disabled` 状态。如果当前没有可以改变状态的绑定，则会在日志中输出一条内容为 `The memory usage of all available bindings exceeds the cache's mem quota. As a result, all available bindings cannot be held on the cache. Please increase the system variable 'tidb_mem_quota_binding_cache' and execute 'admin reload bindings' to ensure that all bindings are available normally` 的警告日志进行提示。需要注意的是当绑定的状态被设置成 `Disabled` 状态的时候，该绑定不会被查询语句所使用。

Reviewer 2: ```suggestion
使用时，只能将 `Disabled` 的绑定改为 `Enabled` 状态或将 `Enabled` 的绑定改为 `Disabled` 状态。如果没有可以改变状态的绑定，则会输出一条警告日志：`The memory usage of all available bindings exceeds the cache's mem quota. As a result, all available bindings cannot be held on the cache. Please increase the system variable 'tidb_mem_quota_binding_cache' and execute 'admin reload bindings' to ensure that all bindings are available normally`。需要注意的是，当绑定被设置成 `Disabled` 状态时，查询语句不会使用该绑定。
```

> 该语句可以将绑定的状态设置成 `Enabled` 或者 `Disabled` 状态，其中只有 `Disabled` 状态的绑定可以被设置成 `Enabled` 状态，只有 `Enabled` 状态的绑定可以被设置成 `Disabled` 状态。如果当前没有可以改变状态的绑定，则会在日志中输出一条内容为 `The memory usage of all available bindings exceeds the cache's mem quota. As a result, all available bindings cannot be held on the cache. Please increase the system variable 'tidb_mem_quota_binding_cache' and execute 'admin reload bindings' to ensure that all bindings are available normally` 的警告日志进行提示。需要注意的是当绑定的状态被设置成 `Disabled` 状态的时候，该绑定不会被查询语句所使用。

Reviewer 3: ````suggestion
该语句可以将绑定的状态设置成 `Enabled` 或者 `Disabled` 状态。注意，该语句只能将 `Disabled` 状态的绑定设置成 `Enabled` 状态，不能将 `Enable` 状态的绑定设置成 `Disabled` 状态。如果当前没有可以改变状态的绑定，则会在日志中输出一条内容如下的的警告日志进行提示：

```
The memory usage of all available bindings exceeds the cache's mem quota. As a result, all available bindings cannot be held on the cache. Please increase the system variable 'tidb_mem_quota_binding_cache' and execute 'admin reload bindings' to ensure that all bindings are available normally`
```

需要注意的是当绑定的状态被设置成 `Disabled` 状态的时候，该绑定不会被查询语句所使用。
````

| original_sql  |  参数化后的原始 SQL |
| bind_sql | 带 Hint 的绑定 SQL |
| default_db | 默认数据库名 |

> | status | 状态，包括 enabled（可用，从 V6.0 的版本开始取代之前版本的 using 状态）、disabled（不可用）、deleted（已删除）、 invalid（无效）、rejected（演进时被拒绝）和 pending verify（等待演进验证） |

Reviewer 2: ```suggestion
| status | 状态，包括 enabled（可用，从 v6.0 开始取代之前版本的 using 状态）、disabled（不可用）、deleted（已删除）、 invalid（无效）、rejected（演进时被拒绝）和 pending verify（等待演进验证） |
```

SELECT * FROM t;
SELECT @@[SESSION.]last_plan_from_binding;
```

```sql
+--------------------------+
| @@last_plan_from_binding |
+--------------------------+
|                        1 |
+--------------------------+
1 row in set (0.00 sec)
```

2. 使用 `explain format = 'verbose'` 语句查看 SQL 语句的查询计划。如果 SQL 语句使用了 binding，`explain` 语句会输出 warning 警告。此时可以通过查看 warning 了解该 SQL 语句使用了哪一条 binding。

```sql
-- 创建一个 global binding

create global binding for
    select * from t
using
    SELECT /*+ USE_INDEX(t, idx_a) */ * FROM t;

-- 使用 explain format = 'verbose' 语句查看 SQL 的执行计划，通过查看 warning 信息确认查询所使用的 binding

explain format = 'verbose' select * from t;
show warnings;
```

```sql
+-------+------+--------------------------------------------------------------------------+
| Level | Code | Message                                                                  |
+-------+------+--------------------------------------------------------------------------+
| Note  | 1105 | Using the bindSQL: SELECT /*+ USE_INDEX(`t` `idx_a`)*/ * FROM `test`.`t` |
+-------+------+--------------------------------------------------------------------------+
1 row in set (0.01 sec)

```

> ### 缓存绑定

Reviewer 2: ```suggestion
### 对绑定进行缓存
```

> ### 绑定的缓存

Reviewer 2: ```suggestion
### 缓存绑定
```

> 在每个 TiDB 实例上都有一个 LRU Cache 对绑定进行缓存，该缓存的容量由系统变量 [tidb_mem_quota_binding_cache](/system-variables#tidb_mem_quota_binding_cache) 进行控制。缓存会影响绑定的使用和查看，在使用和查看绑定的时候，只能使用和查看那些存在于缓存中的绑定。当缓存的容量不足以存放所有的绑定时，系统无法保证缓存中绑定的内容，并且系统会在日志中增加警告日志进行提示。在这种情况下，强烈建议调大缓存的容量，并执行语句 `admin reload bindings` 重新加载系统表中的绑定，以此来保证所有可用的绑定能正常被使用。

Reviewer 3: ```suggestion
在每个 TiDB 实例上都有一个 LRU Cache 对绑定进行缓存，该缓存的容量由系统变量 [`tidb_mem_quota_binding_cache`](/system-variables.md#tidb_mem_quota_binding_cache) 进行控制。缓存会影响绑定的使用和查看。在使用和查看绑定的时候，你只能使用和查看那些存在于缓存中的绑定。当缓存的容量不足以存放所有的绑定时，系统无法保证缓存中绑定的内容，并且系统会在日志中增加警告日志进行提示。在这种情况下，强烈建议调大缓存的容量，并执行语句 `admin reload bindings` 重新加载系统表中的绑定，以此来保证所有可用的绑定能正常被使用。
```

### 排查绑定

绑定的排查通常有两种方式：

> 1. 使用系统变量 [`last_plan_from_binding`](/system-variables.md#last_plan_from_binding-从-v40-版本开始引入) 显示上一条执行语句是否采用 binding 的执行计划。

Reviewer: ```suggestion
- 使用系统变量 [`last_plan_from_binding`](/system-variables.md#last_plan_from_binding-从-v40-版本开始引入) 显示上一条执行语句是否采用 binding 的执行计划。
```

    {{< copyable "sql" >}}

    ```sql
    -- 创建一个 global binding

    CREATE GLOBAL BINDING for
        SELECT * FROM t
    USING
        SELECT /*+ USE_INDEX(t, idx_a) */ * FROM t;

    SELECT * FROM t;
    SELECT @@[SESSION.]last_plan_from_binding;
    ```

    ```sql
    +--------------------------+
    | @@last_plan_from_binding |
    +--------------------------+
    |                        1 |
    +--------------------------+
    1 row in set (0.00 sec)
    ```

> - 使用 `explain format = 'verbose'` 语句查看 SQL 语句的查询计划。如果 SQL 语句使用了 binding，`explain` 语句会输出警告。此时可以通过 `show warnings` 了解该 SQL 语句使用了哪一条 binding。

Reviewer: ```suggestion
- 使用 `explain format = 'verbose'` 语句查看 SQL 语句的查询计划。如果 SQL 语句使用了 binding，可以接着执行 `show warnings` 了解该 SQL 语句使用了哪一条 binding。
```

> 2. 使用 `explain format = 'verbose'` 语句查看 SQL 语句的查询计划。如果 SQL 语句使用了 binding，`explain` 语句会输出 warning 警告。此时可以通过查看 warning 了解该 SQL 语句使用了哪一条 binding。

Reviewer: ```suggestion
- 使用 `explain format = 'verbose'` 语句查看 SQL 语句的查询计划。如果 SQL 语句使用了 binding，`explain` 语句会输出警告。此时可以通过 `show warnings` 了解该 SQL 语句使用了哪一条 binding。
```

    ```sql
    -- 创建一个 global binding

    CREATE GLOBAL BINDING for
        SELECT * FROM t
    USING
        SELECT /*+ USE_INDEX(t, idx_a) */ * FROM t;

    -- 使用 explain format = 'verbose' 语句查看 SQL 的执行计划，通过查看 warning 信息确认查询所使用的 binding

>     explain format = 'verbose' SELECT * FROM t;

Reviewer: 不确定这句后面是漏写了一条警告信息吗

Author: 没有，警告信息是通过 272 行的 show warnings 展示的。展示的内容是 279 行的内容

Reviewer: 252 行提到了 “如果 SQL 语句使用了 binding，`explain` 语句会输出警告。”，就是这个警告不是自动输出的，需要执行 show warnings 手动显示吗？

Author: > 252 行提到了 “如果 SQL 语句使用了 binding，`explain` 语句会输出警告。”，就是这个警告不是自动输出的，需要执行 show warnings 手动显示吗？

对的，是这个意思

Reviewer: 那 L259 可能需要调整下，现在的说法看起来以为explain 语句会自动输出警告

    show warnings;
    ```

    ```sql
    +-------+------+--------------------------------------------------------------------------+
    | Level | Code | Message                                                                  |
    +-------+------+--------------------------------------------------------------------------+
    | Note  | 1105 | Using the bindSQL: SELECT /*+ USE_INDEX(`t` `idx_a`)*/ * FROM `test`.`t` |
    +-------+------+--------------------------------------------------------------------------+
    1 row in set (0.01 sec)

    ```

### 对绑定进行缓存

> 每个 TiDB 实例都有一个 LRU Cache 对绑定进行缓存，缓存的容量由系统变量 [`tidb_mem_quota_binding_cache`](/system-variables.md#tidb_mem_quota_binding_cache从-v60-版本开始引入) 进行控制。缓存会影响绑定的使用和查看，你只能使用和查看存在于缓存中的绑定。

Reviewer: ```suggestion
每个 TiDB 实例都有一个 LRU (Least Recently Used) Cache 对绑定进行缓存，缓存的容量由系统变量 [`tidb_mem_quota_binding_cache`](/system-variables.md#tidb_mem_quota_binding_cache从-v60-版本开始引入) 进行控制。缓存会影响绑定的使用和查看，因此你只能使用和查看存在于缓存中的绑定。
```

> 可以通过执行语句 `show binding_cache status` 来查看绑定的使用情况，该语句无法指定作用域，默认作用域为 GLOBAL。该语句可查看缓存中可用绑定的数量、系统中所有可用绑定的梳理、缓存中所有绑定的内存使用量及缓存的内存容量。

Reviewer: ```suggestion
如需查看绑定的使用情况，可以执行 `SHOW binding_cache status` 语句。该语句无法指定作用域，默认作用域为 GLOBAL。该语句可查看缓存中可用绑定的数量、系统中所有可用绑定的数量、缓存中所有绑定的内存使用量及缓存的内存容量。
```

{{< copyable "sql" >}}

```sql

SHOW binding_cache status;
```

```sql
+-------------------+-------------------+--------------+--------------+
| bindings_in_cache | bindings_in_table | memory_usage | memory_quota |
+-------------------+-------------------+--------------+--------------+
|                 1 |                 1 | 159 Bytes    | 64 MB        |
+-------------------+-------------------+--------------+--------------+
1 row in set (0.00 sec)
```

## 自动捕获绑定 (Baseline Capturing)

> 

Reviewer: 这里能否加一句话，说明什么是自动捕获绑定，有用什么用，什么时候需要开启。

> ### 使用方式
> 

Author: ```suggestion
自动绑定会对符合捕获条件的查询进行捕获，为符合条件的查询生成相应的绑定。通常用于[升级时的计划回退防护](#升级时的计划回退防护)。

### 使用方式

```

### 排查绑定

绑定的排查通常有两种方式：

- 使用系统变量 [`last_plan_from_binding`](/system-variables.md#last_plan_from_binding-从-v40-版本开始引入) 显示上一条执行语句是否采用 binding 的执行计划。

    {{< copyable "sql" >}}

    ```sql
    -- 创建一个 global binding

    CREATE GLOBAL BINDING for
        SELECT * FROM t
    USING
        SELECT /*+ USE_INDEX(t, idx_a) */ * FROM t;

    SELECT * FROM t;
    SELECT @@[SESSION.]last_plan_from_binding;
    ```

    ```sql
    +--------------------------+
    | @@last_plan_from_binding |
    +--------------------------+
    |                        1 |
    +--------------------------+
    1 row in set (0.00 sec)
    ```

- 使用 `explain format = 'verbose'` 语句查看 SQL 语句的查询计划。如果 SQL 语句使用了 binding，可以接着执行 `show warnings` 了解该 SQL 语句使用了哪一条 binding。

    ```sql
    -- 创建一个 global binding

    CREATE GLOBAL BINDING for
        SELECT * FROM t
    USING
        SELECT /*+ USE_INDEX(t, idx_a) */ * FROM t;

>     -- 使用 explain format = 'verbose' 语句查看 SQL 的执行计划，通过查看 warning 信息确认查询所使用的 binding

Author: ```suggestion
    -- 使用 explain format = 'verbose' 语句查看 SQL 的执行计划
```

    explain format = 'verbose' SELECT * FROM t;

>     show warnings;

Author: ```suggestion

    -- 通过执行 `show warnings` 了解该 SQL 语句使用了哪一条 binding

    show warnings;
```

- EXPLAIN 和 EXPLAIN ANALYZE 语句；
- TiDB 内部执行的 SQL 语句，比如统计信息自动加载使用的 SELECT 查询；
- 已经存在 `Enabled`, `Disabled` 状态的绑定对应的语句；

> - 满足绑定黑名单过滤条件的语句；

Reviewer 4: 需要再单独增加一部分说明一些 hint 无法被自动捕获住，比如 JOIN ORDER, MPP 等；

对于 `PREPARE`/`EXECUTE` 语句组，或通过二进制协议执行的查询，TiDB 会为真正的查询（而不是 `PREPARE`/`EXECUTE` 语句）自动捕获绑定。

> **注意：**
>
> 由于 TiDB 存在一些内嵌 SQL 保证一些功能的正确性，所以自动捕获绑定时会默认屏蔽内嵌 SQL。

> ### 绑定过滤黑名单

Reviewer 3: 这部分没有内容，是有待添加吗？

- EXPLAIN 和 EXPLAIN ANALYZE 语句；
- TiDB 内部执行的 SQL 语句，比如统计信息自动加载使用的 SELECT 查询；
- 已经存在 `Enabled`、`Disabled` 状态的绑定对应的语句；
- 满足捕获绑定黑名单过滤条件的语句；

> **注意：**
>

> > 当前的绑定通过生成一组 Hints 来实现对计划的固定，从而确保绑定后的查询语句生成的计划不发生变化。但是受限于当前 Hints 种类，在一些情况下，TiDB 无法保证计划在绑定前后完全一致，其中包括 Join Order、PointGet 和 MPP 相关的计划。

Reviewer 4: `PointGet` 去掉吧，只要 AccessPath 选对，PointGet 一般不会有问题；

> > 当前绑定是通过生成一组 Hints 来实现对计划的固定，从而确保绑定后的查询语句生成的计划能够不发生变化。但是受限于当前 Hints 种类的局限，导致在一些情况下，无法保证计划在绑定前后完全一致，其中包括 Join Order, PointGet 和 MPP 相关的计划。

Reviewer 3: ```suggestion
> 当前的绑定通过生成一组 Hints 来实现对计划的固定，从而确保绑定后的查询语句生成的计划不发生变化。但是受限于当前 Hints 种类，在一些情况下，TiDB 无法保证计划在绑定前后完全一致，其中包括 Join Order、PointGet 和 MPP 相关的计划。
```

对于 `PREPARE`/`EXECUTE` 语句组，或通过二进制协议执行的查询，TiDB 会为真正的查询（而不是 `PREPARE`/`EXECUTE` 语句）自动捕获绑定。

> **注意：**
>
> 由于 TiDB 存在一些内嵌 SQL 保证一些功能的正确性，所以自动捕获绑定时会默认屏蔽内嵌 SQL。

> ### 捕获绑定过滤黑名单

Reviewer 3: ```suggestion
### 捕获绑定的过滤黑名单
```

> 捕获绑定过滤黑名单支持在捕获的过程中排除对满足黑名单规则查询的捕获。黑名单支持的[过滤维度](#过滤维度)包括对表名、频率和用户名进行过滤。

Reviewer 2: ```suggestion
使用本功能，你可以过滤满足黑名单规则的查询，将其排除在捕获范围之外。黑名单支持的[过滤维度](#过滤维度)包括表名、频率和用户名。
```

> 捕获绑定过滤黑名单支持在捕获的过程中排除对满足黑名单规则查询的捕获。黑名单支持的[过滤维度](#过滤维度)包括表名，频率和用户名进行过滤。

Reviewer 3: ```suggestion
捕获绑定过滤黑名单支持在捕获的过程中排除对满足黑名单规则查询的捕获。黑名单支持的[过滤维度](#过滤维度)包括对表名、频率和用户名进行过滤。
```

#### 过滤维度

> | **过滤维度** | **维度名称** | **简介**                                                     | 注意事项                                                     |

Reviewer 2: ```suggestion
| **过滤维度** | **维度名称** | **说明**                                                     | 注意事项                                                     |
```

| :----------- | :----------- | :----------------------------------------------------------- | ------------------------------------------------------------ |

> | 表名         | table        | 按照表名进行过滤，每个过滤规则均采用 `db.table` 形式，支持通配符。详细规则可以参考[直接使用表名](#直接使用表名)和[使用通配符](#使用通配符)。 | 字母大小写不敏感，当插入了非法内容会在日志中输出 `[sql-bind] failed to load mysql.capture_plan_baselines_blacklist` 警告日志进行提示。 |
> | 频率         | frequency    | 默认执行超过一次的语句会被捕获。可以通过插入更大频率的值来变更对捕获语句执行频率的要求。 | 插入小于 1 的值会被认为是非法值，并会在日志中输出 `[sql-bind] frequency threshold is less than 1, ignore it` 警告日志进行提示。同时如果插入了多条变更频率，会选取频率最大的值的作为过滤条件进行过滤。 |
> | 用户名       | user         | 被加入到过滤黑名单的用户名所执行的语句不会被捕获。           | 如果存在一条语句被多个用户名所执行，只有当执行该语句所有的用户名都在黑名单的时候，该语句才不会被捕获。 |

Reviewer 2: ```suggestion
| 表名         | table        | 按照表名进行过滤，每个过滤规则均采用 `db.table` 形式，支持通配符。详细规则可以参考[直接使用表名](#直接使用表名)和[使用通配符](#使用通配符)。 | 字母大小写不敏感，如果包含非法内容，日志会输出 `[sql-bind] failed to load mysql.capture_plan_baselines_blacklist` 警告。 |
| 频率         | frequency    | 默认捕获执行超过一次的语句。可以设置较大值来增加捕获语句的频率。 | 插入的值小于 1 会被认为是非法值，同时，日志会输出 `[sql-bind] frequency threshold is less than 1, ignore it` 警告。如果插入了多条变更频率，频率最大的值会被用作过滤条件。 |
| 用户名       | user         |  黑名单用户名执行的语句不会被捕获。           | 如果多个用户执行同一条语句，只有当他们的用户名都在黑名单的时候，该语句才不会被捕获。 |
```

> **注意：**
>

> > 以下操作都需要数据库的 super privilege 权限。当插入了非法的过滤类型时，会在日志中输出 `[sql-bind] unknown capture filter type, ignore it` 警告日志进行提示。

Reviewer 4: ```suggestion
> 修改黑名单需要数据库的 super privilege 权限。如果黑名单包含了非法的过滤内容时，会在日志中输出 `[sql-bind] unknown capture filter type, ignore it` 进行提示。
```

> > 以下操作都需要数据库的 super privilege 权限。当插入了非法的过滤类型时，会在日志中输出 `[sql-bind] unknown capture filter type, ignore it` 警告日志进行提示。

Reviewer 2: ”以下操作“ 看不出是什么操作。另外，这个 note 的两句话交代了两件事情，结构需要调整。

#### 使用方法

> 将相应的过滤规则插入到系统表 `mysql.capture_plan_baselines_blacklist` 中，在整个集群范围内执行下一次捕获操作时都能生效。

Reviewer 2: ```suggestion
将过滤规则插入到系统表 `mysql.capture_plan_baselines_blacklist` 中，执行下一次捕获时，该过滤规则会在整个集群范围内生效。
```

{{< copyable "sql" >}}

```sql
-- 按照表名进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('table', 'test.t')

> -- 通过正则表达式来实现按照数据库名进行过滤

Reviewer 4: ```suggestion
-- 通过通配符来实现按照数据库名进行过滤
```

Reviewer 4: 这边统一用通配符(wildcard)吧，和正则表达式(regular expression)还不一样；

> INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('table', 'mysql.*')

Reviewer 4: ```suggestion
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('table', 'test.table_*')
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('table', 'db_*.table_*')
```

-- 按照执行频率进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('frequency', '2')

-- 按照用户名进行过滤

> INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('user', 'root')

Reviewer 4: ```suggestion
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('user', 'user1')
```

> insert into mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('table', 'test.t')
> 
> -- 通过正则表达式来实现按照数据库名进行过滤
> insert into mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('table', 'mysql.*')
> 
> -- 按照执行频率进行过滤
> insert into mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('frequency', '2')
> 
> -- 按照用户名进行过滤
> insert into mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('user', 'root')

Reviewer 2: ```suggestion
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('table', 'test.t')

-- 通过正则表达式来实现按照数据库名进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('table', 'mysql.*')

-- 按照执行频率进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('frequency', '2')

-- 按照用户名进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('user', 'root')
```

```

### 验证自动捕获的绑定

> 自动捕获通常被用于确保升级前后执行计划没有回退。当在升级后需要对捕获的绑定进行验证时，可以先通过[变更绑定状态](#变更绑定状态)为 `Disabled` 状态，然后再对相应的查询执行 `Explain` 语句检查其执行计划是否符合预期，如果执行计划符合预期则可以将捕获得到的绑定进行删除。如果执行计划不符合预期，则可以通过[变更绑定状态](#变更绑定状态)为 `Enabled` 状态来保证执行计划不发生回退。

Reviewer 4: 要不直接把这一段加到前面 `变更绑定状态` 小节去吧，这个是变更绑定状态的目的，然后上面有变更绑定状态的用法；

> 自动捕获通常被用于确保升级前后执行计划没有回退。当在升级后需要对捕获的绑定进行验证时，可以先通过[变更绑定状态](#变更绑定状态)为 `Disabled` 状态，然后再对相应的查询执行 `Explain` 语句检查其执行计划是否符合预期，如果执行计划符合预期则可以将捕获得到的绑定进行删除。如果执行计划不符合预期，则可以通过[变更绑定状态](#变更绑定状态)为 `Enabled` 状态来保证执行计划不发生回退。

Reviewer 2: ```suggestion
升级后，可以使用自动捕获检查执行计划的状态，确保不发生回退。

#### 使用方法

1. 根据[变更绑定状态](#变更绑定状态)，将目标绑定变更为 `Disabled`。

2. 对目标查询执行 `Explain` 语句，检查其执行计划是否符合预期。

    - 符合预期，则可以删除捕获的绑定。
    - 不符合预期，则可以[变更绑定状态](#变更绑定状态)为 `Enabled`，确保执行计划不发生回退。
```

- EXPLAIN 和 EXPLAIN ANALYZE 语句；
- TiDB 内部执行的 SQL 语句，比如统计信息自动加载使用的 SELECT 查询；

> - `Enabled` 和 `Disabled` 状态绑定的语句；

Author: ```suggestion
- 存在 `Enabled` 或 `Disabled` 状态绑定的语句；
```

- 满足捕获绑定黑名单过滤条件的语句。

> **注意：**
>

> > 当前的绑定通过生成一组 Hints 来实现对计划的固定，从而确保绑定后的查询语句生成的计划不发生变化。对于大多数 TP 查询，TiDB 能够保证计划前后一致，如使用相同的索引、相同的 Join 方式（如 HashJoin、IndexJoin）等。但是，受限于当前 Hints 的完善程度，一些较为复杂的查询，如两个表以上的 Join 和复杂的 AP、MPP 类查询，TiDB 无法保证计划在绑定前后完全一致。

Reviewer: ```suggestion
> 当前的绑定通过生成一组 Hints 来实现对计划的固定，从而确保绑定后的查询语句生成的计划不发生变化。对于大多数 OLTP 查询，TiDB 能够保证计划前后一致，如使用相同的索引、相同的 Join 方式（如 HashJoin、IndexJoin）等。但是，受限于当前 Hints 的完善程度，对于一些较为复杂的查询，如两个表以上的 Join 和复杂的 OLAP、MPP 类查询，TiDB 无法保证计划在绑定前后完全一致。
```

对于 `PREPARE`/`EXECUTE` 语句组，或通过二进制协议执行的查询，TiDB 会为真正的查询（而不是 `PREPARE`/`EXECUTE` 语句）自动捕获绑定。

> **注意：**
>
> 由于 TiDB 存在一些内嵌 SQL 保证一些功能的正确性，所以自动捕获绑定时会默认屏蔽内嵌 SQL。

### 过滤捕获绑定

> 使用本功能，你可以设置黑名单，将满足黑名单规则的查询排除在捕获范围之外。黑名单支持的[过滤维度](#过滤维度)包括表名、频率和用户名。

Reviewer 2: ```suggestion
使用本功能，你可以设置黑名单，将满足黑名单规则的查询排除在捕获范围之外。黑名单支持的过滤维度包括表名、频率和用户名。
```

#### 使用方式

> 将过滤规则插入到系统表 `mysql.capture_plan_baselines_blacklist` 中，执行下一次捕获时，该过滤规则会在整个集群范围内生效。

Author: ```suggestion
将过滤规则插入到系统表 `mysql.capture_plan_baselines_blacklist` 中，该过滤规则即刻起会在整个集群范围内生效。
```

{{< copyable "sql" >}}

```sql
-- 按照表名进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('table', 'test.t');

-- 通过通配符来实现按照数据库名进行过滤

> INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('table', 'test.table_*');

Reviewer: 这一条是“通过通配符来实现按照表名进行过滤”吗，如果是可以单独注释下吗

Author: 351，352 都是，通配符具体的含义在 363 行里面有个其他的文档有具体解释他的定义。。这边不太好做具体定义上的解释

> -- 通过通配符来实现按照数据库名进行过滤
> INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('table', 'test.table_*');
> INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('table', 'db_*.table_*');

Reviewer: ```suggestion
-- 通过通配符来实现按照数据库名和表名进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('table', 'test.table_*');
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('table', 'db_*.table_*');
```

-- 按照执行频率进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('frequency', '2');

-- 按照用户名进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('user', 'user1');
```

| **过滤维度** | **维度名称** | **说明**                                                     | 注意事项                                                     |
| :----------- | :----------- | :----------------------------------------------------------- | ------------------------------------------------------------ |
| 表名         | table        | 按照表名进行过滤，每个过滤规则均采用 `db.table` 形式，支持通配符。详细规则可以参考[直接使用表名](/table-filter.md#直接使用表名)和[使用通配符](/table-filter.md#使用通配符)。 | 字母大小写不敏感，如果包含非法内容，日志会输出 `[sql-bind] failed to load mysql.capture_plan_baselines_blacklist` 警告。 |

> | 频率         | frequency    | 默认捕获执行超过一次的语句。可以设置较大值来增加捕获语句的频率。 | 插入的值小于 1 会被认为是非法值，同时，日志会输出 `[sql-bind] frequency threshold is less than 1, ignore it` 警告。如果插入了多条变更频率，频率最大的值会被用作过滤条件。 |

Reviewer: “如果插入了多条变更频率” 的意思是用了 VALUES('frequency', '2', '3') 这样的情况吗？如果是，为了便于用户更好地理解，可以加一个例子在它后面吗？

Author: 插入了多条指的是 `VALUES('frequency', '2'), ('frequency', '3')` 这样，加一个例子指的是加在这段话后面，还是后面专门弄一个展示的部分添加相应的例子呢。感觉都不太好操作的样子

Reviewer: 插入了多条指的是 INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('frequency', '2'), ('frequency', '3') 这样吗？ 还是分别执行两次 insert 呢

Author: 两种是一样的，就是插入了两条数据。不管是一次 insert 两个值，还是两次各自 insert 了一个值

Reviewer: 明白了，谢谢，我提了修改建议，将 “如果插入了多条变更频率” 改成了 “如果插入了多条频率过滤规则”，PTAL

| 用户名       | user         | 黑名单用户名执行的语句不会被捕获。                           | 如果多个用户执行同一条语句，只有当他们的用户名都在黑名单的时候，该语句才不会被捕获。 |

> **注意：**
>
> - 修改黑名单需要数据库的 super privilege 权限。
>

> > - 如果黑名单包含了非法的过滤内容时，会在日志中输出 `[sql-bind] unknown capture filter type, ignore it` 进行提示。

Reviewer: ```suggestion
> - 如果黑名单包含了非法的过滤内容时，TiDB 会在日志中输出 `[sql-bind] unknown capture filter type, ignore it` 进行提示。
```

### 升级时的计划回退防护

> 利用自动捕获绑定，可以在 TiDB 集群升级时，对潜在的计划回退风险起到一定程度的防护作用，具体流程为：

Reviewer: ```suggestion
当需要升级 TiDB 集群时，你可以利用自动捕获绑定对潜在的计划回退风险进行一定程度的防护，具体流程为：
```

1. 升级前打开自动捕获一段时间，让大多数重要的计划（出现过两次及以上）都能被捕获到。

    > **注意：**
    >
    > 经测试，长期打开自动捕获对集群负载的性能影响很小。尽量长期打开自动捕获，以确保重要的查询都能被捕获到。

2. 进行 TiDB 集群的升级。在升级完成后，这些通过捕获的绑定会发挥作用，确保在升级后，查询的计划不会改变。
3. 升级完成后，根据情况手动删除绑定。

>     - 检查绑定来源：

Reviewer: 请问用户可以如何”检查绑定来源“ 呢？有对应的 SQL 语句可以输出看到`Source` 字段吗

Author: 208 行 `查看绑定` 这部分有对如何查看绑定相关的信息做具体介绍，感觉应该不需要在这边在额外说明一下了。

Reviewer: 是的，不用单独说明，加一个链接就行。我在下面的 comment 里加上了，PTAL.

>     - 检查绑定来源：

Reviewer: ```suggestion
    - 通过[`SHOW BINDINGS`](#查看绑定)语句检查绑定来源：
```

>         根据 `Source` 字段对绑定的来源进行区分，确认是 (`capture`) 得到还是通过手动创建(`manual`)。

Author: ```suggestion
        根据 `Source` 字段对绑定的来源进行区分，确认是通过捕获(`capture`) 生成还是通过手动创建(`manual`) 生成。
```

>         根据 `Source` 字段对绑定的来源进行区分，确认是通过捕获(`capture`) 生成还是通过手动创建(`manual`) 生成。

Reviewer: ```suggestion
        根据 `Source` 字段对绑定的来源进行区分，确认是通过捕获 (`capture`)  生成还是通过手动创建 (`manual`) 生成。
```

>         根据 `Source` 字段对绑定的来源进行区分，确认是通过捕获(`capture`) 生成还是通过手动创建(`manual`) 生成。

Reviewer: ```suggestion
        根据输出中的 `Source` 字段对绑定的来源进行区分，确认是通过捕获(`capture`) 生成还是通过手动创建(`manual`) 生成。
```

    - 确定 `capture` 的绑定是否需要保留：

        ```
        -- 查看绑定生效时的计划
        SET @@SESSION.TIDB_USE_PLAN_BASELINES = true;
        EXPLAIN FORMAT='VERBOSE' SELECT * FROM t1 WHERE ...;

        -- 查看绑定不生效时的计划
        SET @@SESSION.TIDB_USE_PLAN_BASELINES = false;
        EXPLAIN FORMAT='VERBOSE' SELECT * FROM t1 WHERE ...;
        ```

        - 如果屏蔽绑定前后，查询得到的计划一致，则可以安全删除此绑定。

>         - 如果计划不一样，则可能需要对此计划变化的原因进行排查，如检查统计信息等操作，在这种情况下需要保留此绑定，确保计划不发生变化。

Reviewer: ```suggestion
        - 如果计划不一样，则可能需要对此计划变化的原因进行排查，如检查统计信息等操作。在这种情况下需要保留此绑定，确保计划不发生变化。
```

| :----------- | :----------- | :----------------------------------------------------------- | ------------------------------------------------------------ |
| 表名         | table        | 按照表名进行过滤，每个过滤规则均采用 `db.table` 形式，支持通配符。详细规则可以参考[直接使用表名](/table-filter.md#直接使用表名)和[使用通配符](/table-filter.md#使用通配符)。 | 字母大小写不敏感，如果包含非法内容，日志会输出 `[sql-bind] failed to load mysql.capture_plan_baselines_blacklist` 警告。 |
| 频率         | frequency    | 默认捕获执行超过一次的语句。可以设置较大值来增加捕获语句的频率。 | 插入的值小于 1 会被认为是非法值，同时，日志会输出 `[sql-bind] frequency threshold is less than 1, ignore it` 警告。如果插入了多条变更频率，频率最大的值会被用作过滤条件。 |
| 用户名       | user         | 黑名单用户名执行的语句不会被捕获。                           | 如果多个用户执行同一条语句，只有当他们的用户名都在黑名单的时候，该语句才不会被捕获。 |

> **注意：**
>
> - 修改黑名单需要数据库的 super privilege 权限。
>
> - 如果黑名单包含了非法的过滤内容时，TiDB 会在日志中输出 `[sql-bind] unknown capture filter type, ignore it` 进行提示。

### 升级时的计划回退防护

当需要升级 TiDB 集群时，你可以利用自动捕获绑定对潜在的计划回退风险进行一定程度的防护，具体流程为：

1. 升级前打开自动捕获一段时间，让大多数重要的计划（出现过两次及以上）都能被捕获到。

    > **注意：**
    >
    > 经测试，长期打开自动捕获对集群负载的性能影响很小。尽量长期打开自动捕获，以确保重要的查询都能被捕获到。

2. 进行 TiDB 集群的升级。在升级完成后，这些通过捕获的绑定会发挥作用，确保在升级后，查询的计划不会改变。
3. 升级完成后，根据情况手动删除绑定。

    - 通过[`SHOW BINDINGS`](#查看绑定)语句检查绑定来源：

>         根据输出中的 `Source` 字段对绑定的来源进行区分，确认是通过捕获(`capture`) 生成还是通过手动创建(`manual`) 生成。

Author: ```suggestion
        根据输出中的 `Source` 字段对绑定的来源进行区分，确认是通过捕获 (`capture`) 生成还是通过手动创建 (`manual`) 生成。
```

{{< copyable "sql" >}}

```sql
-- 按照表名进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('table', 'test.t');

-- 通过通配符来实现按照数据库名和表名进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('table', 'test.table_*');
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('table', 'db_*.table_*');

-- 按照执行频率进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('frequency', '2');

-- 按照用户名进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('user', 'user1');
```

| **过滤维度** | **维度名称** | **说明**                                                     | 注意事项                                                     |
| :----------- | :----------- | :----------------------------------------------------------- | ------------------------------------------------------------ |
| 表名         | table        | 按照表名进行过滤，每个过滤规则均采用 `db.table` 形式，支持通配符。详细规则可以参考[直接使用表名](/table-filter.md#直接使用表名)和[使用通配符](/table-filter.md#使用通配符)。 | 字母大小写不敏感，如果包含非法内容，日志会输出 `[sql-bind] failed to load mysql.capture_plan_baselines_blacklist` 警告。 |

> | 频率         | frequency    | 默认捕获执行超过一次的语句。可以设置较大值来增加捕获语句的频率。 | 插入的值小于 1 会被认为是非法值，同时，日志会输出 `[sql-bind] frequency threshold is less than 1, ignore it` 警告。如果插入了多条变更频率，频率最大的值会被用作过滤条件。 |

Reviewer: ```suggestion
| 频率         | frequency    | 默认捕获执行超过一次的语句。可以设置较大值来增加捕获语句的频率。 | 插入的值小于 1 会被认为是非法值，同时，日志会输出 `[sql-bind] frequency threshold is less than 1, ignore it` 警告。如果插入了多条频率过滤规则，频率最大的值会被用作过滤条件。 |
```

> **注意：**
>
> 如果黑名单包含了非法的过滤内容时，会在日志中输出 `[sql-bind] unknown capture filter type, ignore it` 进行提示。

#### 使用方法

将过滤规则插入到系统表 `mysql.capture_plan_baselines_blacklist` 中，执行下一次捕获时，该过滤规则会在整个集群范围内生效。

> **注意：**
>
> 修改黑名单需要数据库的 super privilege 权限。

{{< copyable "sql" >}}

```sql
-- 按照表名进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('table', 'test.t')

-- 通过通配符来实现按照数据库名进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('table', 'test.table_*')
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('table', 'db_*.table_*')

-- 按照执行频率进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('frequency', '2')

-- 按照用户名进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('user', 'user1')
```

### 验证自动捕获的绑定

升级后，可以使用自动捕获检查执行计划的状态，确保不发生执行计划回退。

#### 使用方法

1. 创建一个单独的 Session 进行验证，在该 Session 上设置 `set @@session.tidb_use_plan_baselines = false`，使得在当前 Session 上的绑定不生效。

2. 逐一检查通过捕获得到的绑定，对绑定中的目标查询执行 `Explain` 语句，检查其执行计划是否符合预期。

    - 符合预期，则可以删除通过捕获得到的绑定。
    - 不符合预期，则保留对应的绑定，确保执行计划不发生回退。

> 

Reviewer 4: 这里要不直接给用户以 “升级时的计划回退防护” 为例子进行讲解吧：

### 升级时的计划回退防护

利用自动捕获绑定，可以在 TiDB 升级时，对计划回退进行一定程度的防护，具体流程为：

1. 升级前在一台 TiDB 上打开自动捕获一段时间，让它捕获大多数重要的计划（出现过两次及以上）；

2. 进行 TiDB 升级，升级过程中，这些捕获的绑定会发挥作用，确保在升级过程中，计划不会改变；

3. 升级完成后，根据情况手动对这些绑定进行删除；

在对绑定进行删除时，可以根据 `Source` 字段进行区分，这些绑定是自动捕获(`capture`)的还是手动创建(`manual`)。

在删除某个绑定前，建议检查下这个绑定是否正在发挥作用，避免删除后计划发生改变，具体的检查方式为：

```sql
-- 查看在绑定生效情况下的计划
EXPLAIN FORMAT='VERBOSE' SELECT * FROM t1 WHERE ...

-- 查看绑定不能生效情况下的计划，可通过设置变量 TIDB_USE_PLAN_BASELINES 来屏蔽掉绑定的影响
SET @@SESSION.TIDB_USE_PLAN_BASELINES = false
EXPLAIN FORMAT='VERBOSE' SELECT * FROM t1 WHERE ...
```

如果屏蔽绑定前后，计划一样，则可以安全删除此绑定；如果计划不一样，则可能需要对此计划变化进行排查，如检查统计信息等，或者手动创建一条绑定来固定计划。

- EXPLAIN 和 EXPLAIN ANALYZE 语句；
- TiDB 内部执行的 SQL 语句，比如统计信息自动加载使用的 SELECT 查询；
- 已经存在 `Enabled`、`Disabled` 状态的绑定对应的语句；
- 满足捕获绑定黑名单过滤条件的语句；

> **注意：**
>

> > 当前的绑定通过生成一组 Hints 来实现对计划的固定，从而确保绑定后的查询语句生成的计划不发生变化。但是受限于当前 Hints 的完善程度，一些较为复杂的查询，如 2 表以上的 Join 和复杂的 AP、MPP 类查询，TiDB 无法保证计划在绑定前后完全一致。对于大多数 TP 查询，TiDB 能够保证计划前后一致，如使用相同的索引、相同的 Join 方式（如 HashJoin、IndexJoin）等。

Reviewer 2: ```suggestion
> 当前的绑定通过生成一组 Hints 来实现对计划的固定，从而确保绑定后的查询语句生成的计划不发生变化。对于大多数 TP 查询，TiDB 能够保证计划前后一致，如使用相同的索引、相同的 Join 方式（如 HashJoin、IndexJoin）等。但是，受限于当前 Hints 的完善程度，一些较为复杂的查询，如两个表以上的 Join 和复杂的 AP、MPP 类查询，TiDB 无法保证计划在绑定前后完全一致。
```
1. 先说能实现的，再说不能实现的。
2. 2 表是指两个表么？

对于 `PREPARE`/`EXECUTE` 语句组，或通过二进制协议执行的查询，TiDB 会为真正的查询（而不是 `PREPARE`/`EXECUTE` 语句）自动捕获绑定。

> **注意：**
>
> 由于 TiDB 存在一些内嵌 SQL 保证一些功能的正确性，所以自动捕获绑定时会默认屏蔽内嵌 SQL。

### 过滤捕获绑定

> 使用本功能，你可以过滤满足黑名单规则的查询，将其排除在捕获范围之外。黑名单支持的[过滤维度](#过滤维度)包括表名、频率和用户名。

Reviewer 2: ```suggestion
使用本功能，你可以设置黑名单，将满足黑名单规则的查询排除在捕获范围之外。黑名单支持的[过滤维度](#过滤维度)包括表名、频率和用户名。
```

> ### 捕获绑定过滤黑名单
> 
> 使用本功能，你可以过滤满足黑名单规则的查询，将其排除在捕获范围之外。黑名单支持的[过滤维度](#过滤维度)包括表名、频率和用户名。

Author: ```suggestion
### 过滤捕获绑定

使用本功能，你可以过滤满足黑名单规则的查询，将其排除在捕获范围之外。黑名单支持的[过滤维度](#过滤维度)包括表名、频率和用户名。
```

> #### 过滤维度

Reviewer 2: ```suggestion
#### 设置过滤黑名单
```

| **过滤维度** | **维度名称** | **说明**                                                     | 注意事项                                                     |
| :----------- | :----------- | :----------------------------------------------------------- | ------------------------------------------------------------ |
| 表名         | table        | 按照表名进行过滤，每个过滤规则均采用 `db.table` 形式，支持通配符。详细规则可以参考[直接使用表名](/table-filter.md#直接使用表名)和[使用通配符](/table-filter.md#使用通配符)。 | 字母大小写不敏感，如果包含非法内容，日志会输出 `[sql-bind] failed to load mysql.capture_plan_baselines_blacklist` 警告。 |
| 频率         | frequency    | 默认捕获执行超过一次的语句。可以设置较大值来增加捕获语句的频率。 | 插入的值小于 1 会被认为是非法值，同时，日志会输出 `[sql-bind] frequency threshold is less than 1, ignore it` 警告。如果插入了多条变更频率，频率最大的值会被用作过滤条件。 |
| 用户名       | user         | 黑名单用户名执行的语句不会被捕获。                           | 如果多个用户执行同一条语句，只有当他们的用户名都在黑名单的时候，该语句才不会被捕获。 |

> **注意：**
>
> 如果黑名单包含了非法的过滤内容时，会在日志中输出 `[sql-bind] unknown capture filter type, ignore it` 进行提示。

> #### 使用方法

Reviewer 2: ```suggestion
#### 使过滤生效
```

将过滤规则插入到系统表 `mysql.capture_plan_baselines_blacklist` 中，执行下一次捕获时，该过滤规则会在整个集群范围内生效。

> **注意：**
>
> 修改黑名单需要数据库的 super privilege 权限。

{{< copyable "sql" >}}

```sql
-- 按照表名进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('table', 'test.t');

-- 通过通配符来实现按照数据库名进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('table', 'test.table_*');
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('table', 'db_*.table_*');

-- 按照执行频率进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('frequency', '2');

-- 按照用户名进行过滤
INSERT INTO mysql.capture_plan_baselines_blacklist(filter_type, filter_value) VALUES('user', 'user1');
```

### 升级时的计划回退防护

利用自动捕获绑定，可以在 TiDB 集群升级时，对潜在的计划回退风险起到一定程度的防护作用，具体流程为：

1. 升级前打开自动捕获一段时间，让大多数重要的计划（出现过两次及以上）都能被捕获到；
2. 进行 TiDB 集群的升级。在升级完成后，这些通过捕获的绑定会发挥作用，确保在升级后，查询的计划不会改变；
3. 升级完成后，根据情况手动对这些绑定进行删除；

> **注意：**
>
> 经过测试发现，长期打开自动捕获的开关对集群负载的性能影响很小。打开自动捕获开关的时间最好足够久，确保重要的查询都能被捕获到。

> 在对绑定进行删除时，可以根据 `Source` 字段对绑定的来源进行区分，是通过捕获(`capture`) 得到的还是通过手动创建(`manual`) 得到的。

Reviewer 2: ```suggestion
删除绑定时，可以根据 `Source` 字段区分绑定的来源，可能为捕获(`capture`)，也可能为手动创建(`manual`)。
```
区分这个有什么用意？

    ```sql
    -- 在待升级的版本上检查自动演进的开关 `tidb_evolve_plan_baselines` 是否关闭。

> 

Author: The editor automatically formats this part of the content. Although it is a useless change, can it be adopted if it conforms to the specification?

Reviewer 3: LGTM
